### PR TITLE
#479 infinite scroll report page

### PR DESCRIPTION
**Pull request by yunfangjuan**
# Description
Render more rows on the report page if necessary

## Type of change

- [X ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce
Outline the steps to test or reproduce the PR here.

1. Go to http://localhost:3000/samples/18 
2. scroll and try out filters. 

## Impacted Areas in Application
List general components of the application that this PR will affect:

Which pages?

- [ ] Login Screen
- [ ] Single Upload Page
- [ ] Batch Upload Page
- [ ] All Projects Page
- [ ] Sample Details Page
- [X ] Report Page

Specific components 

* PipelineSampleReport.jsx

# Testing Script:

###Login Page
* Successfully login to the application and get taken to the project page

###Single Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Uploading a duplicate file throws an error
* Pipeline gets initiated successfully 

###Batch Upload
* Using an AWS path that doesn't exist throws an error
* Succesfully uploaded files from a folder (use test host)
* Navigate to batch upload settings page after upload 
* Host selection from previous page still selected 
* Uploading a duplicate file throws a warning
* Pipeline gets initiated successfully 

###All Project Page
* Switch between projects 
* Search for a sample
* Navigate through different pages of samples
* Select Upload new sample button
* Select sample to see generated report

###Sample Details Page
* Information is consistent from Project Page
* Notes are editable and save successfully 
* Can navigate to report page 

###Report Page
* Data loads succesfully 
* Collapse and Uncollapse button works
* Clicking on the taxonomy name gives you total reads for that taxon
* Category filter works
* Genus Search works
* Filters work as expected
* Disabling Filters works
* Filtering by column works 

# Checklist:

- [ ] I have run through the testing script to make sure current functionality is unchanged
- [ ] I have done relevant tests that prove my fix is effective or that my feature works
- [ ] I have spent time testing out edge cases for my feature
- [ ] I have updated the test script or pull request template if necessary
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/assets/src/components/PipelineSampleReport.jsx [hunk 1]
```diff
@@ -208,10 +224,48 @@ class PipelineSampleReport extends React.Component {
...
+      this.setState({rendering: false})
+    }
+    );
+  }
+
+  //Load more samples on scroll
+  scrollDown() {
+    var that = this;
+    $(window).scroll(function() {
+      if ($(window).scrollTop() > $(document).height() - $(window).height() - 6000) {
```
Comment (boris-dimitrov) [inline app/assets/src/components/PipelineSampleReport.jsx]: 6000?

Comment (boris-dimitrov) [reply app/assets/src/components/PipelineSampleReport.jsx]: Does this mean you've scrolled past >6000 invisible items?

Comment (yunfangjuan) [reply app/assets/src/components/PipelineSampleReport.jsx]: 6000 pixels. basically every time you scroll ;-)

File: app/assets/src/components/PipelineSampleReport.jsx [hunk 2]
```diff
@@ -208,10 +224,48 @@ class PipelineSampleReport extends React.Component {
...
+        {that.state.rows_total > 0 && !that.state.rendering ? that.renderMore() : null }
```
Comment (boris-dimitrov) [inline app/assets/src/components/PipelineSampleReport.jsx]: Is that an `if` statement condensed to one line for stylistic reasons?  Or does "? :"  have other special meaning in JS?

File: app/assets/src/components/PipelineSampleReport.jsx [hunk 3]
```diff
@@ -208,10 +224,48 @@ class PipelineSampleReport extends React.Component {
...
 
+
+  renderMore() {
+    let selected_taxons = this.state.selected_taxons
+    let currentPage = this.state.pagesRendered
+    let rowsPerPage = this.max_rows_to_render
+    if (selected_taxons.length > currentPage * this.max_rows_to_render) {
+      let next_page = selected_taxons.slice(currentPage * rowsPerPage, rowsPerPage * (currentPage +1))
+      this.setState((prevState) => ({
+        selected_taxons_top: [...prevState.selected_taxons_top, ...next_page],
```
Comment (boris-dimitrov) [inline app/assets/src/components/PipelineSampleReport.jsx]: so theoretically as you scroll down each render will take longer and longer?

**Comment by boris-dimitrov** [comment]
This was very interesting, thanks.

File: app/assets/src/components/PipelineSampleReport.jsx (hunk 3 above)
Comment (yunfangjuan) [reply app/assets/src/components/PipelineSampleReport.jsx]: yep.

**Comment by yunfangjuan** [comment]
@trevorb reviewed it and gave my good suggestions. I will update this diff soon.

**Comment by trevorb** [comment]
Just to record what @yunfangjuan chatted about.  I suggested rather than having to remember to use setState callbacks everywhere in the component to keep the rendering flag accurate that we use the componentWillUpdate and componentDidUpdate lifecycle events.

**Pull request merged**